### PR TITLE
Shutdown gracefully when recieving a termination signal

### LINKED ISSUE #218: Stopping jellyfin container may result in sqlite DB remaining locked
After stopping my jellyfin container using `docker-compose stop jellyfin`, pulling the latest image and running it using `docker-compose run -d jellyfin` I discovered that the container wasn't starting correctly but instead continually restarting.

The error message JF shows on startup is the following:
 ```
UnhandledException
Busy: database is locked
SQLitePCL.pretty.SQLiteException: Exception of type 'SQLitePCL.pretty.SQLiteException' was thrown.
   at SQLitePCL.pretty.SQLiteException.CheckOk(sqlite3 db, Int32 rc)
   at SQLitePCL.pretty.SQLiteDatabaseConnection.PrepareStatement(String sql, String& tail)
   at SQLitePCL.pretty.DatabaseConnection.PrepareStatement(IDatabaseConnection This, String sql)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Emby.Server.Implementations.Data.BaseSqliteRepository.CreateConnection(Boolean isReadOnly) in /repo/Emby.Server.Implementations/Data/BaseSqliteRepository.cs:line 113
   at Emby.Server.Implementations.Data.SqliteItemRepository.Initialize(SqliteUserDataRepository userDataRepo, IUserManager userManager) in /repo/Emby.Server.Implementations/Data/SqliteItemRepository.cs:line 125
   at Emby.Server.Implementations.ApplicationHost.RegisterResources() in /repo/Emby.Server.Implementations/ApplicationHost.cs:line 856
   at Emby.Server.Implementations.ApplicationHost.Init() in /repo/Emby.Server.Implementations/ApplicationHost.cs:line 817
   at MediaBrowser.Server.Mono.MainClass.RunApplication(ServerApplicationPaths appPaths, ILogManager logManager, StartupOptions options)
   at MediaBrowser.Server.Mono.MainClass.Main(String[] args)
SQLitePCL.pretty.SQLiteException
   at SQLitePCL.pretty.SQLiteException.CheckOk(sqlite3 db, Int32 rc)
   at SQLitePCL.pretty.SQLiteDatabaseConnection.PrepareStatement(String sql, String& tail)
   at SQLitePCL.pretty.DatabaseConnection.PrepareStatement(IDatabaseConnection This, String sql)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Emby.Server.Implementations.Data.BaseSqliteRepository.CreateConnection(Boolean isReadOnly) in /repo/Emby.Server.Implementations/Data/BaseSqliteRepository.cs:line 113
   at Emby.Server.Implementations.Data.SqliteItemRepository.Initialize(SqliteUserDataRepository userDataRepo, IUserManager userManager) in /repo/Emby.Server.Implementations/Data/SqliteItemRepository.cs:line 125
   at Emby.Server.Implementations.ApplicationHost.RegisterResources() in /repo/Emby.Server.Implementations/ApplicationHost.cs:line 856
   at Emby.Server.Implementations.ApplicationHost.Init() in /repo/Emby.Server.Implementations/ApplicationHost.cs:line 817
   at MediaBrowser.Server.Mono.MainClass.RunApplication(ServerApplicationPaths appPaths, ILogManager logManager, StartupOptions options)
   at MediaBrowser.Server.Mono.MainClass.Main(String[] args)
```

To me this looks like the SQLite DB wasn't unlocked properly when Jellyfin was shutdown.

I'm still investigating whether this issue could be caused by something on my setup in particular (e.g. file permissions, although unlikely) but it could also very well be an rare race condition that happens whenever jellyfin is shutdown during a DB write.

### LINKED ISSUE #422: Database not being written to in docker containers
For me and at least one other, at least the users.db is not being written to properly inside of docker containers.

Opening the users.db file results in an empty database UNLESS you copy the cache files users.db-shm and users.db-whl as well.

I am really curious as to whether or not this behavior is consistent with other methods of running Jellyfin as this is really not good.

### LINKED ISSUE #454: Checkpoint database files on service stop
As a developer, it would be nice if the database files were checkpointed on service stop, so the .db files reflected the system at shutdown.

Jellyfin uses Write-Ahead Logging instead of a rollback journal for its database approach.  Stopping Jellyfin via the command line (sudo service jellyfin stop) does not checkpoint the databases, leaving the -wal and -shm files present.  Shutdown via the GUI does checkpoint.

### PR DESCRIPTION
**Pull request by Bond-009**
**Changes**
Graceful shutdown on SIGTERM and Ctrl+C

**Issues**
Fixes #218
Fixes #422 
Fixes #454

### CODE REVIEW

**Comment by Bond-009** [comment]
Don't merge yet!
Ctrl+C breaks some stuff

File: Jellyfin.Server/Program.cs [hunk 1]
```diff
@@ -1,313 +1,335 @@
...
+            // $JELLYFIN_LOG_DIR needs to be set for the logger configuration manager
+            Environment.SetEnvironmentVariable("JELLYFIN_LOG_DIR", appPaths.LogDirectoryPath);
+            await createLogger(appPaths);
+            _logger = _loggerFactory.CreateLogger("Main");
+
+            AppDomain.CurrentDomain.UnhandledException += (sender, e)
+                => _logger.LogCritical((Exception)e.ExceptionObject, "Unhandled Exception");
+
+            // Intercept Ctrl+C and Ctrl+Break
+            Console.CancelKeyPress += (sender, e) =>
```
Comment (EraYaN) [inline Jellyfin.Server/Program.cs]: Does this also catch SIGINT when sent from a kill command for example? (Or an orchestrator)
A tty might not be connected.

Comment (Bond-009) [reply Jellyfin.Server/Program.cs]:
```
Occurs when the System.ConsoleModifiers.Control modifier key (Ctrl) and either the System.ConsoleKey.C console key (C) or the Break key are pressed simultaneously (Ctrl+C or Ctrl+Break).
```
Doesn't look like it.

Comment (hawken93) [reply Jellyfin.Server/Program.cs]: I've done some digging and windows needs key handlers while linux needs signal handling. Is it possible to do both? If there is no way to register SIGINT, I bet that the linux implementation will catch SIGINT and emulate Ctrl+C to the application. Don't quote me on this though.

**Comment by hawken93** [comment]
LGTM if that matters :)

File: Jellyfin.Server/Program.cs (hunk 1 above)
Comment (EraYaN) [reply Jellyfin.Server/Program.cs]: And that seems to not be in dotnet core anyway. Otherwise we could do something with unsafe code.

**Comment by ploughpuff** [comment]
This change also fixed #454.

```
root@vb-deb-jelly:/home/ploughpuff# ls -la /var/lib/jellyfin/data/
total 868
drwxr-xr-x 5 jellyfin jellyfin   4096 Jan 14 13:13 .
drwxr-xr-x 8 jellyfin jellyfin   4096 Jan 14 13:13 ..
-rw-r--r-- 1 jellyfin jellyfin   8192 Jan 14 13:13 activitylog.db
-rw-r--r-- 1 jellyfin jellyfin  32768 Jan 14 13:13 activitylog.db-shm
-rw-r--r-- 1 jellyfin jellyfin   8272 Jan 14 13:13 activitylog.db-wal
-rw-r--r-- 1 jellyfin jellyfin  28672 Jan 14 13:13 authentication.db
-rw-r--r-- 1 jellyfin jellyfin  32768 Jan 14 13:13 authentication.db-shm
-rw-r--r-- 1 jellyfin jellyfin  12392 Jan 14 13:13 authentication.db-wal
drwxr-xr-x 5 jellyfin jellyfin   4096 Jan 14 13:13 devices
-rw-r--r-- 1 jellyfin jellyfin     35 Jan 14 13:13 device.txt
-rw-r--r-- 1 jellyfin jellyfin  12288 Jan 14 13:13 displaypreferences.db
-rw-r--r-- 1 jellyfin jellyfin  32768 Jan 14 13:13 displaypreferences.db-shm
-rw-r--r-- 1 jellyfin jellyfin      0 Jan 14 13:13 displaypreferences.db-wal
-rw-r--r-- 1 jellyfin jellyfin 163840 Jan 14 13:13 library.db
-rw-r--r-- 1 jellyfin jellyfin  32768 Jan 14 13:14 library.db-shm
-rw-r--r-- 1 jellyfin jellyfin 436752 Jan 14 13:13 library.db-wal
drwxr-xr-x 2 jellyfin jellyfin   4096 Jan 14 13:13 playlists
drwxr-xr-x 2 jellyfin jellyfin   4096 Jan 14 13:13 ScheduledTasks
-rw-r--r-- 1 jellyfin jellyfin   8192 Jan 14 13:13 users.db
-rw-r--r-- 1 jellyfin jellyfin  32768 Jan 14 13:13 users.db-shm
-rw-r--r-- 1 jellyfin jellyfin   8272 Jan 14 13:13 users.db-wal
root@vb-deb-jelly:/home/ploughpuff# service jellyfin stop
root@vb-deb-jelly:/home/ploughpuff# ls -la /var/lib/jellyfin/data/
total 240
drwxr-xr-x 5 jellyfin jellyfin   4096 Jan 14 13:15 .
drwxr-xr-x 8 jellyfin jellyfin   4096 Jan 14 13:13 ..
-rw-r--r-- 1 jellyfin jellyfin   8192 Jan 14 13:15 activitylog.db
-rw-r--r-- 1 jellyfin jellyfin  28672 Jan 14 13:15 authentication.db
drwxr-xr-x 5 jellyfin jellyfin   4096 Jan 14 13:13 devices
-rw-r--r-- 1 jellyfin jellyfin     35 Jan 14 13:13 device.txt
-rw-r--r-- 1 jellyfin jellyfin  12288 Jan 14 13:13 displaypreferences.db
-rw-r--r-- 1 jellyfin jellyfin 163840 Jan 14 13:15 library.db
drwxr-xr-x 2 jellyfin jellyfin   4096 Jan 14 13:13 playlists
drwxr-xr-x 2 jellyfin jellyfin   4096 Jan 14 13:13 ScheduledTasks
-rw-r--r-- 1 jellyfin jellyfin   8192 Jan 14 13:15 users.db

```

Good job, Bond+2 :1st_place_medal:

**Pull request merged**